Add support for switching distribution for all integration tests

alpar-t opened the pull request:
Allow using the `tests.distribution` property for telling integration tests which distribution to use. 
If the property is not present, integration tests  for plugins will use `integ-test-zip`, qa integration tests will default to `oss-zip`, and x-pack qa to `zip`.
When the property is present, integration tests will always honor it, meaning that integration tests will be skipped if they cannot run against the specific distribution. e.x. we won't run `x-pack/qa` with `-Dtests.distribution=oss-zip` because we can't honor that, ignoring the property and running them with `zip` would just duplicate the work. 

TODO:

- [x] make use of  `tests.distribution` for plugins to switch distribution these run against
- [x] add an info level log to mention which distribution was used when starting up the cluster 
- [x] skip  the `x-pack/plugins/build.gradle` ( top level ) integration tests if distribution is not `zip` 
- [x] skip `xpack-qa` tests if `tests.distribution` and it's not `zip`
- [x] what happens if we pass `integ-test-zip` to QA ? Tests should probably be skipped. 
- [x] test the whole build without `tests.distribution` and with each of `zip`, `oss-zip`
- [x] fix plugin and module tests so they work with different dsitributions

alpar-t committed:
remove left-over comment

alpar-t committed:
make sure of the property for plugins

elasticmachine commented:
Pinging @elastic/es-core-infra

atorok commented:
@nik9000 @rjernst did I start down the right path ?

nik9000 commented on plugins/discovery-file/build.gradle:
```diff
@@ -38,7 +38,7 @@ task setupSeedNodeAndUnicastHostsFile(type: DefaultTask) {
 // for unicast discovery
 ClusterConfiguration config = new ClusterConfiguration(project)
-config.distribution = 'integ-test-zip'
+config.distribution = System.getProperty('tests.distribution', 'integ-test-zip')
```
I'm not sure we need this line at all given your change to `PluginBuildPlugin`.

atorok replied:
This would take precedence and override that. So it would be set back to `integ-test-zip` regardless of the property.

rjernst replied:
I am concerned this is fragile. I don't want to see this logic copied around to other projects. Can you please create an issue for making creating additional cluster fixtures easier, so that the default can be set by PluginBuildPlugin for this project and reused the we create this cluster?

nik9000 replied:
++ on this being fragile. Fixing it in a followup is fine with me though.

nik9000 commented on qa/smoke-test-rank-eval-with-mustache/build.gradle:
```diff
@@ -24,10 +24,4 @@ apply plugin: 'elasticsearch.rest-test'
     testCompile project(path: ':modules:lang-mustache', configuration: 'runtime')
-}
-
-/*
```
++. This is closed. If you want to just push a commit that removes that comment to the repo without waiting on a PR that'd be awesome.

nik9000 commented on plugins/examples/painless-whitelist/build.gradle:
```diff
@@ -30,8 +30,12 @@ dependencies {
 
-integTestCluster {
-  distribution = 'zip'
+if (System.getProperty('tests.distribution') == null) {
```
This seems a bit complicated to add to every project. I'm not 100% sure we need to support `-Dtests.distribution=integ-test-zip`. Maybe `integ-test-zip` is a thing we should only use when you don't override the property.

nik9000 replied:
It'd be fairly simple to fail the build if someone sets `distribution` to `integ-test-zip`. If we did that then this project could do `distribution = System.getProperty('tests.distribution', 'zip)`.

nik9000 replied:
I *might* be cleaner to move the `PluginBuildPlugin` change to a `gradle.projectsEvaluated` and have it handle the logic all in one place.

atorok replied:
setting distro `integ-test-zip` would offer a way to run some integration tests in a build that would be much quicker as it would skip a lot of heavy lifting. I'm not sure how useful that would be during development, so can't decide if it's worth doing.

rjernst replied:
I don't think that is generally useful. Plugins and modules are the projects which use integ-test-zip, and you can already cd into the plugins or modules dir and run tests from there.

nik9000 replied:
Yeah. I don't think we should support setting `-Dtests.distribution=integ-test-zip`. It should be what you get when you run without setting it.

atorok commented:
@nik9000 just want to make sure that you looked at the todo list as well and share your thoughts before I start implementing

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy:
```diff
@@ -532,9 +532,6 @@ class ClusterFormationTasks {
     }
 
     static Task configureInstallModuleTask(String name, Project project, Task setup, NodeInfo node, Project module) {
-        if (node.config.distribution != 'integ-test-zip') {
```
If using a non integ-test-zip distribution, modules need to not be installed at all. This looks like it allows installing them over what the distribution installs, which could create craziness. I think we need to skip setting up install module tasks if the distribution is not integ-test-zip.

rjernst reviewed (commented):
I have some general comments

alpar-t committed:
skip installing modules if these exist in the distribution

alpar-t committed:
Log the distrbution being ran

alpar-t committed:
Don't allow running with integ-tests-zip passed externally

alpar-t committed:
top level x-pack/qa can't run with oss distro

alpar-t committed:
Merge branch 'master' into test/distributions

atorok commented:
@nik9000 @rjernst I adressed hopefully all comments. opened #30903 and #30904.

alpar-t committed:
Add support for matching objects in lists

Makes it possible to have a key that points to a list and assert that a
certain object is present in the list. All keys have to be present and
values have to match. The objects in the source list may have additional
fields.

example:
```
  match:  { 'nodes.$master.plugins': { name: ingest-attachment }  }
```

alpar-t committed:
Update plugin and module tests to work with other distributions

Some of the tests expected that the integration tests will always be ran
with  the `integ-test-zip` distribution so that there will be no other
plugins loaded.

With this change, we check for the presence of the plugin without
assuming exclusivity.

alpar-t committed:
Allow modules to run on other distros as well

To match the behavior of tets.distributions

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java:
```diff
@@ -81,7 +90,36 @@ protected void doAssert(Object actualValue, Object expectedValue) {
+        if(expectedValue instanceof Map && actualValue instanceof List) {
+            Map<String, Object> expectedMap = (Map<String, Object>) expectedValue;
+            List<Object> actualList = (List<Object>) actualValue;
+            assertTrue(
```
I'm not sure we want `match` to be `contains` in some cases. If we want this we should make a new construct.

atorok replied:
The semantics seemed to fit with me, but now that I think of it, `match` is in fact an exact match so it should be an exact match when dealing with lists as well. I'll add a new construct and call it `contains`.

alpar-t committed:
Add and use a new `contains` assertion

Replaces the  previus changes that caused `match` to do a partial match.

alpar-t committed:
Merge commit '8e4ab82' into test/distributions

atorok commented:
test this please

atorok commented:
@nik9000 can you take another look please.

alpar-t committed:
Merge commit 'fcf1e41' into test/distributions

alpar-t committed:
Merge remote-tracking branch 'origin/master' into test/distributions

nik9000 commented on modules/lang-expression/src/test/resources/rest-api-spec/test/lang_expression/10_basic.yml:
```diff
@@ -1,4 +1,4 @@
-# Integration tests for Expression scripts
+# Integration tests for Expression scriptsmatch
```
Did you mean to add this?

atorok replied:
no, I'll have to be more careful with my fat fingers :) thanks !

nik9000 commented on plugins/examples/painless-whitelist/build.gradle:
```diff
@@ -30,8 +30,12 @@ dependencies {
+  integTestCluster.distribution = 'oss-zip'
+}
+
+if (integTestCluster.distribution == 'integ-test-zip') {
```
I think we can remove this line because we don't support setting the setting to this.

nik9000 replied:
We can add something to cluster formation tasks to make sure we don't set the system property to `integ-test-zip`.

nik9000 commented on qa/build.gradle:
```diff
@@ -5,6 +5,9 @@ subprojects { Project subproj ->
   subproj.tasks.withType(RestIntegTestTask) {
     subproj.extensions.configure("${it.name}Cluster") { cluster ->
       cluster.distribution = System.getProperty('tests.distribution', 'oss-zip')
+      if (cluster.distribution == 'integ-test-zip') {
```
I think we're better off checking if the system property is `integ-test-zip` in the cluster formation tasks and failing there. Basically, the only way you can get `integ-test-zip` is to not set the property.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/ContainsAssertion.java:
```diff
@@ -0,0 +1,80 @@
+            Map<String, Object> expectedMap = (Map<String, Object>) expectedValue;
+            List<Object> actualList = (List<Object>) actualValue;
+            List<Map<String, Object>> actualValues = actualList.stream()
+                .filter(each -> each instanceof Map)
```
I think we should fail the assertion if the thing we're testing isn't a map.

atorok replied:
As implemented now, we can check that a list has an object that has all the keys at the specific values.  In theory we could have a list  having a mix of lists and objects. The filter only rules out anything that's not a map at this point when looking for `actualValues`. The subsequent check on actual values not empty would fail the assertion in case of a list of lists for example. 
The only case we don't fail if we have a list that has both a list and an object that has the keys we are interested in, but I don't think we should in this case, because we _did_ find that the list contains what we are looking for.

nik9000 replied:
I see! Got it.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/MatchAssertion.java:
```diff
@@ -26,13 +26,22 @@
+import static org.hamcrest.Matchers.not;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
```
These look like leftovers.

nik9000 commented on x-pack/plugin/src/test/resources/rest-api-spec/test/xpack/10_basic.yml:
```diff
@@ -1,4 +1,4 @@
-# Integration tests for monitoring
+# Integration tests for monitoring.23.
```
Mistake?

nik9000 commented on x-pack/plugin/src/test/resources/rest-api-spec/test/xpack/10_basic.yml:
```diff
@@ -10,14 +10,14 @@
-    - match:  { nodes.$master.modules.21.name: x-pack-sql  }
-    - match:  { nodes.$master.modules.22.name: x-pack-upgrade  }
-    - match:  { nodes.$master.modules.23.name: x-pack-watcher  }
+    - contains:  { nodes.$master.modules: { name: x-pack-core } }
```
This seems nice!

nik9000 reviewed (changes requested):
I think it this is still a good thing to do. I've left a few requests to change things. Thanks for pinging me again on it!

alpar-t committed:
Merge commit '86423f9' into test/distributions

alpar-t committed:
Implement PR review comments

nik9000 reviewed (approved):
Thanks for working with me on this one! LGTM.
